Makefile

```diff
@@ -15,68 +15,80 @@
 # specific language governing permissions and limitations
 # under the License.
 
+
+.PHONY: all \
+        runtime vta cpptest crttest \
+        lint pylint cpplint scalalint \
+	doc \
+	web webclean \
+	cython cython3 cyclean \
+        clean
+
+# Remember the root directory, to be usable by submake invocation.
 ROOTDIR = $(CURDIR)
-# Specify an alternate output directory relative to ROOTDIR. Default build
-OUTPUTDIR = $(if $(OUTDIR), $(OUTDIR), build)
 
-.PHONY: clean all test doc pylint cpplint scalalint lint\
-	 cython cython2 cython3 web runtime vta
+# Specify an alternate output directory relative to ROOTDIR.  Defaults
+# to "build".  Can also be a space-separated list of build
+# directories, each with a different configuation.
+TVM_BUILD_PATH ?= build
+TVM_BUILD_PATH := $(abspath $(TVM_BUILD_PATH))
 
-ifndef DMLC_CORE_PATH
-  DMLC_CORE_PATH = $(ROOTDIR)/3rdparty/dmlc-core
-endif
+# Allow environment variables for 3rd-party libraries, default to
+# packaged version.
+DMLC_CORE_PATH ?= $(ROOTDIR)/3rdparty/dmlc-core
+DLPACK_PATH ?= $(ROOTDIR)/3rdparty/dlpack
+VTA_HW_PATH ?= $(ROOTDIR)/3rdparty/vta-hw
 
-ifndef DLPACK_PATH
-  DLPACK_PATH = $(ROOTDIR)/3rdparty/dlpack
-endif
 
-ifndef VTA_HW_PATH
-  VTA_HW_PATH = $(ROOTDIR)/3rdparty/vta-hw
-endif
 
-INCLUDE_FLAGS = -Iinclude -I$(DLPACK_PATH)/include -I$(DMLC_CORE_PATH)/include
-PKG_CFLAGS = -std=c++11 -Wall -O2 $(INCLUDE_FLAGS) -fPIC
-PKG_LDFLAGS =
 
+all: $(addsuffix /all,$(TVM_BUILD_PATH))
 
-all:
-	@mkdir -p $(OUTPUTDIR) && cd $(OUTPUTDIR) && cmake .. && $(MAKE)
+runtime: $(addsuffix /runtime,$(TVM_BUILD_PATH))
+vta: $(addsuffix /vta,$(TVM_BUILD_PATH))
+cpptest: $(addsuffix /cpptest,$(TVM_BUILD_PATH))
+crttest: $(addsuffix /crttest,$(TVM_BUILD_PATH))
 
-runtime:
-	@mkdir -p $(OUTPUTDIR) && cd $(OUTPUTDIR) && cmake .. && $(MAKE) runtime
+# Set up a default config.cmake inside the build directory.
+# filter-out used to avoid circular dependency.
+%/config.cmake: | $(filter-out %/config.cmake,$(ROOTDIR)/cmake/config.cmake)
```

Comment:
This changes the current behavior where a config.cmake in the root can also serve the purpose of configuration. Given that the cmake rule already looks for the config.cmake, perhaps we can skip this step

Reply:
Good point, I had been going off of the "Install from Source" documentation, and hadn't realized there could be a root directory config.cmake instead.  I'd like to have the `%/config.cmake` build rule in place so that it can later be specialized to set up local build directories to match the CI.  (e.g. The rule for `build_docker/ci_gpu/config.cmake` would copy `config.cmake` to that location, then run `tests/scripts/task_config_build_gpu.sh` on it.)

I think the best way to maintain the current behavior while allowing for this later expansion would be to symlink to a root directory `config.cmake` if it exists, falling back to copying the `config.cmake` if it doesn't.

Reply:
The latest commit on this PR now preserves the behavior of using the root-directory's `config.cmake` using symlinks.

Reply:
symlink is likely not necessary, because the cmake is already made to use that behavior https://github.com/apache/tvm/blob/main/CMakeLists.txt#L16, as long as no config.cmake is copied it is fine

Reply:
True, and testing without the symlink shows that it works correctly.  I tend to be hesitant on makefile rules that neither create their target nor are marked as `.PHONY`, but it does have the desired behavior here.  We may need the symlink at some point in the future, if we want to avoid the call to `cmake` when possible, but that can be done at a later point.

Reply:
Right now makefile is mainly served as a convenient tool and we are quite cmake centric, so we can keep things simple in the Makefile

Reply:
That definitely makes sense, this is primarily me looking to add more convenience where possible.